src/librustc_target/abi/call/mod.rs

```diff
@@ -284,21 +318,34 @@ impl<'a, Ty> TyLayout<'a, Ty> {
 
                 for i in 0..self.fields.count() {
                     if !is_union && total != self.fields.offset(i) {
-                        return None;
+                        return HomogeneousAggregate::Heterogeneous;
                     }
 
                     let field = self.field(cx, i);
+
+                    // Skip over zero-sized fields for the purposes of
+                    // this check, since they don't represent any
+                    // data.
```

Comment:
ah hmm removing this line made some of the union tests fail, not sure why, will check in a bit

Reply:
fixed, I was handling zero-length arrays in unions wrong